Bluetooth: controller: Avoid sizeof to access ULL/LLL struct member

cvinayak opened the pull request:
Avoid using sizeof to access ULL and LLL struct members.
Based on the alignment requirements of structures, due to
padding between structure members, use of sizeof of previous
struct member to access next struct member is incorrect.

Continue to use explicitly stored parent pointer to access
ULL context. Combine event header and ULL header so that
the parent pointer point directly to the combined ULL
struct.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

wopu-ot commented:
I assume that this relates to the `sizeof` in `HDR_ULL2LLL`. As far as I can tell, the actual types are known in all places where this macro is used, so it should be possible to eliminate the use of this macro by proper casting. I think that would be a much better solution than relying on how the structures are exactly laid out. Or am I overlooking something?

cvinayak commented:
>  it should be possible to eliminate the use of this macro by proper casting

@wopu-ot Let me give a second try, will replace `HDR_ULL`  and `HDR_ULL2LLL` with a macro that uses `offsetof` with case-by-case specific ULL types.

cvinayak committed:
Bluetooth: controller: Avoid sizeof to access ULL/LLL struct member

Avoid using sizeof to access ULL and LLL struct members.
Based on the alignment requirements of structures, due to
padding between structure members, use of sizeof of previous
struct member to access next struct member is incorrect.

Continue to use explicitly stored parent pointer to access
ULL context. Combine event header and ULL header so that
the parent pointer point directly to the combined ULL
struct.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Fix connection terminate to happen on event done

Fix conn_cleanup to be performed in the event done. The
regression was introduced in
commit 5412f61ac61a ("Bluetooth: controller: Introduce
separate done memq").

Now that done event is handled asynchronously outside the
handling of the terminate node rx type, it is necessary that
terminate node rx type is generated after the done event.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Fix auxiliary scan context release

Fix the auxiliary scan context release to be performed in
the disabled_cb callback after the ULL reference count is
decremented.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

mtpr-ot reviewed (dismissed):
Nice cleanup and regression fix. Approved.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_conn_iso.c:
```diff
@@ -181,15 +181,14 @@ void ull_conn_iso_cis_established(struct ll_conn_iso_stream *cis)
-	/* Skip if CIG terminated by local host */
-	if (unlikely(lll->handle == 0xFFFF)) {
-		return;
-	}
```
As `conn_cleanup` is now only called when processing done event, same should be followed for connected ISO too, that any cleanup should be initiated only on event done.

This means,

1. if there is a race where in `ticker_cb` has finished prepare and incremented reference count in the ULL context, then after which the done event is doing the cleanup, then this check is redundant as the handle is set to 0xFFFF when the done event is processed.
2. if there is a race where `done` event in the ULL context is being processed for initiating cleanup and `ticker_stop` operation has been enqueued. At this point lets assume `ticker_cb` executes due to ticker timeout happening just after ULL context processed the done event, then this is a case similar to (1) above wherein the event is placed into the pipeline and has the reference count incremented.
3. if the ticker stop....

Wait,  in case 2, pipeline has the event which is not flushed. Ok, the removed checks have to be back so that event does not proceed after the handle has been set to 0xFFFF.

Thank you for making me re-analyse the race. I will revert back these deleted code.

mtpr-ot replied:
Glad to help - would have been nice to get rid of them.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -1512,20 +1515,16 @@ int ull_disable(void *lll)
-	if (!hdr) {
-		return ULL_STATUS_SUCCESS;
+	hdr = HDR_LLL2ULL(lll);
+	if (!hdr || !ull_ref_get(hdr)) {
```
The reference count is decremented when "done" event has been processed in the ULL context, and which means "done" event which is only generated when the radio event is completed and has released radio resources. This means LLL of a particular state/role is no longer using the radio, and hence there is no need to call/tigger lll_disable.

Also, as `ull_disable` is called after `ticker_stop` operation has been completed and checked using semaphore, hence no new event would get its reference count incremented while being here.

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -5513,12 +5551,14 @@ static inline void ctrl_tx_ack(struct ll_conn *conn, struct node_tx **tx,
+			conn->llcp_terminate.reason_final =
 			      pdu_tx->llctrl.terminate_ind.error_code;
-
-		conn_cleanup(conn, reason);
```
It is not entirely clear to me where the corresponding call to `conn_cleanup` happens after the change. Is it from `ull_conn_done` (line 1228)?

cvinayak replied:
yes. https://github.com/zephyrproject-rtos/zephyr/pull/33367/files?file-filters%5B%5D=.c#diff-135b49c1c599aa1060145905101574cb58224b39926a1e1a04dda6ff6e8702a0R1232

wopu-ot reviewed (approved):
LGTM

mtpr-ot reviewed (commented):
Review in progress, but our simulation tests are failing with this PR so I need time to investigate.

mtpr-ot commented on subsys/bluetooth/controller/ll_sw/ull_conn_types.h:
```diff
@@ -31,7 +31,6 @@ enum llcp {
 };
 
 struct ll_conn {
-	struct evt_hdr  evt;
```
Removal of this struct makes the `ull_hdr` the first item. Because elements of type `struct ll_conn` reside in a memory pool, the `mem_init` initially uses the first 32-bit as an internal pointer. Previously this was the memory occupied by `evt_hdr->ticks_xtal_to_start`, which is always assigned before use. Now the memory shared with the internal pointer is  `ull_hdr->ref` (and unused padding bytes). This is fatal because the member is never initialized.
The problem can be fixed by setting `hdr->ref = 0` in `ull_hdr_init`.

wopu-ot replied:
Awesome catch!

cvinayak replied:
Good catch. The changes in this PR has disclosed many out of bound and uninitialized memory accesses that existed. Will add the initialization and update this PR.

cvinayak committed:
Bluetooth: controller: Fix uninitialized ULL reference count

Fix missing initialization of reference count in
ull_hdr_init function.

This has not caused issues so far, but when the ref member
of the struct ull_hdr if placed in the beginning of a
context that is allocated using mem_acquire function then
first few bytes used would make the ref member to have
uninitialized value when such context is allocated by
mem_acquire. First few bytes are the next pointer and free
count stored by the mem module.

The issue was discovered in subsequent commits that
restructure the ULL context structures.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

mtpr-ot reviewed (approved):
LGTM
